File: src/ipc/sources.ts
import { call, inTauri } from "./client";
import type { Source } from "./types.gen";

export const sourcesApi = {
  list: (projectId: string) => call<Source[]>("source_list", { projectId }),
  get: (projectId: string, sourceId: string) =>
    call<Source>("source_get", { projectId, sourceId }),
  addFiles: (projectId: string, paths: string[]) =>
    call<Source[]>("source_add_files", { input: { projectId, paths } }),
  addUrl: (projectId: string, url: string) =>
    call<Source>("source_add_url", { projectId, url }),
  addFolder: (projectId: string, folder: string) =>
    call<Source>("source_add_folder", { projectId, folder }),
  reanalyze: (projectId: string, sourceId: string) =>
    call<void>("source_reanalyze", { projectId, sourceId }),
  delete: (projectId: string, sourceId: string) =>
    call<void>("source_delete", { projectId, sourceId }),
};

/** OS file picker for adding sources. Returns absolute paths (never shown to the
 * user directly — the frontend only holds ids after this). */
export async function pickSourceFiles(): Promise<string[]> {
  if (!inTauri) return [];
  const { open } = await import("@tauri-apps/plugin-dialog");
  const picked = await open({
    multiple: true,
    directory: false,
    title: "Add sources",
  });
  if (!picked) return [];
  return Array.isArray(picked) ? picked : [picked];
}

/** OS folder picker for importing a static-site folder (issue 5). */
export async function pickFolder(): Promise<string | null> {
  if (!inTauri) return null;
  const { open } = await import("@tauri-apps/plugin-dialog");
  const picked = await open({ directory: true, multiple: false, title: "Add website folder" });
  return typeof picked === "string" ? picked : null;
}

File: src-tauri/src/commands/sources.rs
use crate::domain::source::*;
use crate::error::AppResult;
use crate::services::{projects, sources};
use crate::state::AppState;
use tauri::{AppHandle, State};

#[tauri::command]
pub fn source_list(state: State<'_, AppState>, project_id: String) -> AppResult<Vec<Source>> {
    let db = projects::open_db(&state.projects_dir, &project_id)?;
    sources::list(&db)
}

#[tauri::command]
pub fn source_get(
    state: State<'_, AppState>,
    project_id: String,
    source_id: String,
) -> AppResult<Source> {
    let db = projects::open_db(&state.projects_dir, &project_id)?;
    sources::get(&db, &source_id)
}

#[tauri::command]
pub fn source_add_files(
    app: AppHandle,
    state: State<'_, AppState>,
    input: AddFilesInput,
) -> AppResult<Vec<Source>> {
    sources::add_files(
        &app,
        &state.app_db_path,
        &state.projects_dir,
        state.jobs.clone(),
        &input.project_id,
        input.paths,
    )
}

#[tauri::command]
pub fn source_add_url(
    app: AppHandle,
    state: State<'_, AppState>,
    project_id: String,
    url: String,
) -> AppResult<crate::domain::source::Source> {
    sources::add_url(
        &app,
        &state.app_db_path,
        &state.projects_dir,
        state.jobs.clone(),
        &project_id,
        &url,
    )
}

#[tauri::command]
pub fn source_add_folder(
    app: AppHandle,
    state: State<'_, AppState>,
    project_id: String,
    folder: String,
) -> AppResult<Source> {
    sources::add_folder(
        &app,
        &state.app_db_path,
        &state.projects_dir,
        state.jobs.clone(),
        &project_id,
        &folder,
    )
}

#[tauri::command]
pub fn source_reanalyze(
    app: AppHandle,
    state: State<'_, AppState>,
    project_id: String,
    source_id: String,
) -> AppResult<()> {
    sources::reanalyze(
        &app,
        &state.app_db_path,
        &state.projects_dir,
        state.jobs.clone(),
        &project_id,
        &source_id,
    )
}

#[tauri::command]
pub fn source_delete(
    state: State<'_, AppState>,
    project_id: String,
    source_id: String,
) -> AppResult<()> {
    state.with_db(|db| sources::delete(db, &state.projects_dir, &project_id, &source_id))
}
